c++/src/kj/async-inl.h

```diff
@@ -217,13 +217,10 @@ class Event: private AsyncObject {
   SourceLocation location;
 };
 
-class PromiseNode: private AsyncObject {
-  // A Promise<T> contains a chain of PromiseNodes tracking the pending transformations.
-  //
-  // To reduce generated code bloat, PromiseNode is not a template.  Instead, it makes very hacky
-  // use of pointers to ExceptionOrValue which actually point to ExceptionOr<T>, but are only
-  // so down-cast in the few places that really need to be templated.  Luckily this is all
-  // internal implementation details.
+class PromiseArenaMember {
+  // An object that is allocated in a PromiseArena. `PromiseNode` inherits this, and most
+  // arena-allocated objects are `PromiseNode` subclasses, but `TaskSet::Task` and potentially
+  // other objects that commonly live on the end of a promise chain can also leverage this.
```

Comment:
Yeah the refcounting makes it difficult... I would probably have to re-implement refcounting rather than use kj::Refcounted. That said, that does seem like an interesting thing to explore... maybe I will later.

Reply:
Went ahead and implemented this. Worth another couple of percentage points!